File: Mat-Mat-BLAS/C/mat-mat-blas.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#include <sys/types.h>
#include <sys/resource.h>
#include <sys/time.h>

#include <mpi.h>

#define  N      1000
#define  DEBUG  1
#define  EPS    1.0e-18

/* Please define here for the matrices. */
static double  A[N][N];
static double  B[N][N];
static double  C[N][N];

void MyMatMat(double [][], double [][], double [][], int); 


void main(int argc, char* argv[]) {

     double  t1, t2, t_w;
     double  dc_inv, d_mflops;

     int     i, j;      
     int     iflag, iflag_t;


     /* matrix generation --------------------------*/
     if (DEBUG == 1) {
       for(j=0; j<N; j++) {
         for(i=0; i<N; i++) {
           A[j][i] = 1.0;
           B[j][i] = 1.0;
           C[j][i] = 0.0;
         }
       }
     } else {
       srand(1);
       dc_inv = 1.0/(double)RAND_MAX;
       for(j=0; j<N; j++) {
         for(i=0; i<N; i++) {
           A[j][i] = rand()*dc_inv;
           B[j][i] = rand()*dc_inv;
           C[j][i] = 0.0;
         }
       }
     } /* end of matrix generation --------------------------*/

     /* Start of mat-mat routine ----------------------------*/
     t1 = MPI_Wtime();

     MyMatMat(C, A, B, N);

     t2 = MPI_Wtime();
     t_w =  t2 - t1; 
     /* End of mat-mat routine --------------------------- */

     printf("N  = %d \n",N);
     printf("Mat-Mat time  = %lf [sec.] \n",t_w);

     d_mflops = 2.0*(double)N*(double)N*(double)N/t_w;
     d_mflops = d_mflops * 1.0e-6;
     printf(" %lf [MFLOPS] \n", d_mflops);

     /* Verification routine ----------------- */
     if (DEBUG == 1) {
       iflag = 0;
       for(j=0; j<N; j++) { 
         for(i=0; i<N; i++) { 
           if (fabs(C[j][i] - (double)N) > EPS) {
             printf(" Error! in ( %d , %d ) th argument. \n",j, i);
             iflag = 1;
             break;
           } 
         }
       }
       if (iflag == 0) printf("OK! \n");
     }
     /* ------------------------------------- */
     printf("nenett\n");
     exit(0);

}

void MyMatMat(double C[N][N], double A[N][N], double B[N][N], int n) 
{
  int  i, j, k;
  double ALPHA, BETA;
  char TEX[1] = {'N'};

  //  for(i=0; i<n; i++) {
  // for(j=0; j<n; j++) {
  //   for(k=0; k<n; k++) {
  //     C[i][j] += A[i][k] * B[k][j]; 
  //   }
  //  }
  //}

  //  Please write BLAS call in here.
  //  dgemm_(....);
  ALPHA = BETA = 1;
  dgemm_(&TEX,&TEX ,&n ,&n ,&n ,&ALPHA ,A ,&n ,B ,&n ,&BETA ,C ,&n );
   
}

File: Mat-Mat-BLAS/F/mat-mat-blas.f
       program main

       include 'mpif.h'
       include 'mat-mat-blas.inc'

       integer DEBUG
       parameter (DEBUG=1)
       double precision EPS
       parameter (EPS=1.0e-18)


       double precision  A(NN, NN)
       double precision  B(NN, NN)
       double precision  C(NN, NN)

       integer N
       double precision  t0, t1, t2
       double precision  d_mflops, dtemp
    
       integer  i, j
       integer  iflag

       N = NN

c      === matrix generation --------------------------
       if (DEBUG .eq. 1) then
          do j=1, N
            do i=1, N
              A(i, j) = 1.0;
              B(i, j) = 1.0;
              C(i, j) = 0.0;
            enddo
          enddo
       else 
          call RANDOM_SEED
          do j=1, N
            do i=1, N
              call RANDOM_NUMBER(dtemp)   
              A(i, j) = dtemp
              call RANDOM_NUMBER(dtemp)   
              B(i, j) = dtemp
              C(i, j) = 0.0
            enddo
          enddo
       endif
c      === end of matrix generation ------------------------

c      === Start of mat-mat routine ----------------------------
       t1 = MPI_Wtime()

       call MyMatMat(C, A, B, N)

       t2 = MPI_Wtime()

       t0 =  t2 - t1 
c      === End of mat-mat routine ---------------------------

       print *, "N  = ", N
       print *, "Mat-Mat time[sec.] = ", t0

       d_mflops = 2.0*dble(N)*dble(N)*dble(N)/t0
       d_mflops = d_mflops * 1.0e-6
       print *, "MFLOPS = ", d_mflops

       if (DEBUG .eq. 1) then
c      === Verification routine ----------------- 
         iflag = 0
         do j=1, N
           do i=1, N 
             if (dabs(C(i,j) - dble(N)) > EPS) then
               print *, " Error! in (", i, ",", j, ") th argument."
               iflag = 1
               goto 10
             endif 
           enddo
         enddo
 10      continue
c        -----------------------------------------

         if (iflag .eq. 0) then
           print *, " OK!"
         endif
      
       endif
c      -----------------------------------------

       stop
       end



       subroutine MyMatMat(C ,A, B, n)
       include 'mat-mat-blas.inc'
       double precision C(NN, NN)
       double precision A(NN, NN)
       double precision B(NN, NN)
       integer n

       integer  i, j, k
       double precision ALPHA,BETA

       do i=1, n
         do j=1, n
           do k=1, n 
             C(i, j) = C(i, j) + A(i, k) * B(k, j) 
           enddo
         enddo
       enddo

c      Please write BLAS call in here.
c      CALL DGEMM(....)


      return
      end
